src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,366 @@
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					&syntax.Regexp{
+						Op:    syntax.OpQuest,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							newRe,
+						},
+					},
+					original,
+				},
+			}, true, nil
+		}
+	case syntax.OpPlus:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					&syntax.Regexp{
+						Op:    syntax.OpStar,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							original.Sub[0],
+						},
+					},
+				},
+			}, true, nil
+		}
+	case syntax.OpRepeat:
+		if len(parsed.Sub) > 0 && parsed.Min > 0 && parsed.Max > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			original.Min--
+			original.Max--
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					original,
+				},
+			}, true, nil
+		}
+	}
+	return parsed, false, nil
+}
+
+func deepCopy(ast *syntax.Regexp) *syntax.Regexp {
+	if ast == nil {
+		return nil
+	}
+	copied := *ast
+	copied.Sub = make([]*syntax.Regexp, 0, len(ast.Sub))
+	for _, r := range ast.Sub {
+		copied.Sub = append(copied.Sub, deepCopy(r))
+	}
+	if len(copied.Sub0) != 0 && copied.Sub0[0] != nil {
+		copied.Sub0[0] = deepCopy(copied.Sub0[0])
+	}
+	// NB(prateek): we don't copy ast.Rune (which could be a heap allocated slice) intentionally,
+	// because none of the transformations we apply modify the Rune slice.
+	return &copied
+}
+
+var emptyStringOps = []syntax.Op{
+	syntax.OpEmptyMatch, syntax.OpQuest, syntax.OpPlus, syntax.OpStar, syntax.OpRepeat,
+}
+
+func matchesEmptyString(ast *syntax.Regexp) bool {
+	if ast == nil {
+		return false
+	}
+	for _, op := range emptyStringOps {
+		if ast.Op == op {
+			if len(ast.Sub) > 0 {
+				return matchesEmptyString(ast.Sub[0])
+			}
+			return true
+		}
+	}
+	return false
+}
+
+func simplify(ast *syntax.Regexp) *syntax.Regexp {
+	newAst, _ := simplifyHelper(ast)
+	return newAst
+}
+
+func simplifyHelper(ast *syntax.Regexp) (*syntax.Regexp, bool) {
+	if ast == nil {
+		return nil, false
+	}
+	switch ast.Op {
+	case syntax.OpConcat:
+		// a concatenation of a single sub-expression is the same as the sub-expression itself
+		if len(ast.Sub) == 1 {
+			return ast.Sub[0], true
+		}
+
+		changed := false
+		// check if we have any concats of concats, if so, we can pull the ones below this level up
+		subs := make([]*syntax.Regexp, 0, len(ast.Sub))
+		for _, sub := range ast.Sub {
+			if sub.Op == syntax.OpConcat {
+				subs = append(subs, sub.Sub...)
```

Comment:
Hm you don't have to take into account each sub expression too?

Reply:
Oh you do this later, I see.

Reply:
Hm but shouldn't you check if each sub expression matches empty string itself (like you do below for the non concat case)?